koku/masu/celery/tasks.py

```diff
@@ -370,3 +372,34 @@ def clean_volume():
     LOG.info("Removing all files older than %s", expiration_date)
     LOG.info("The following files were too new to delete: %s", retain_files)
     LOG.info("The following files were deleted: %s", deleted_files)
+
+
+@app.task(name="masu.celery.tasks.crawl_account_hierarchy", queue_name="crawl_account_hierarchy")
+def crawl_account_hierarchy():
+    """Crawl top level accounts to discover hierarchy."""
+    _, polling_accounts = Orchestrator.get_accounts()
+    LOG.info("Account hierarchy crawler found %s accounts to scan" % len(polling_accounts))
+    processed = 0
+    skipped = 0
+    for account in polling_accounts:
+        crawler = None
+
+        # Look for a known crawler class to handle this provider
+        if account.get("provider_type") == Provider.PROVIDER_AWS:
+            crawler = AWSOrgUnitCrawler(account)
+
+        if crawler:
+            LOG.info(
+                "Starting account hierarchy crawler for type {} with provider_uuid ({})".format(
+                    account.get("provider_type"), account.get("provider_uuid")
+                )
+            )
+            crawler.crawl_account_hierarchy()
```

Comment:
It looks like this is going to do a serial crawl of each account. So, as the number of accounts increase, the time it takes for this to run will increase.

Before we try to optimize that, do you have a sense for how long each account takes to crawl? Like, will we need to worry about this task's runtime at 100 accounts or 10,000 accounts?

At some point, we'll probably need to adjust this, but before we worry about that, I just want to do a quick "gut check" for whether that needs to happen now versus 9-12 months from now.

Reply:
Yeah, during local testing I used the openshift dedicated aws account, which has about 2000 accounts associated with it. It was pretty fast on my mac taking about 3-4 minutes on average to collect the entire hierarchy tree. The benefit of the save method we are using is that we are doing a get_or_create. So the hardest hitting run will always be the first run and all the runs after that is mostly just adding new information. 

That being said, I would be open to any performance suggestions.

Reply:
Some back of the napkin math says that should cover us for a few thousand accounts of similar complexity. So, I think we'll be okay for a while. I see no need to optimize this right now.